---
title: 'STAT 222: Gene Genome: Integrated Genomic Analysis for Glioblastoma'
author: "Cheng Li, Jingyi Tan, Laura Katz"
date: "March 5, 2017"
output: pdf_document
---

## Load Package:
```{r}
library(readr)
library(tidyr)
library(dplyr)
library(reshape2)
```

## Read in data
```{r}
Broad202 = read_delim("222_data/Broad202.txt","\t", escape_double = FALSE, trim_ws = TRUE)
LBL202 = read_delim("222_data/LBL202.txt","\t", escape_double = FALSE, trim_ws = TRUE)
UNC202 = read_delim("222_data/UNC202.txt","\t", escape_double = FALSE, trim_ws = TRUE)
colnames(UNC202)[1]="Gene"

firstrow=read_delim("222_data/unifiedScaled.txt", 
                    "\t", escape_double = FALSE, col_names = FALSE,
                    trim_ws = TRUE, skip = 0,n_max=1)
unifiedScaled = read_delim("222_data/unifiedScaled.txt",
                           "\t", escape_double = FALSE, col_names = FALSE, 
                           trim_ws = TRUE, skip = 1) %>% 
  `colnames<-`(value=c('Gene',firstrow))
rm(list="firstrow")
```

## Reshape the data
```{r}
merge_genes=intersect(LBL202$Gene,UNC202$Gene) %>% intersect(Broad202$Gene)
merge_ids=intersect(colnames(LBL202)[-1],colnames(UNC202)[-1]) %>% 
  intersect(colnames(Broad202)[-1])
unc= UNC202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(UNC202))) %>%
  melt(id.vars="Gene")
broad=Broad202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(Broad202))) %>%
  melt(id.vars='Gene')
lbl=LBL202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(LBL202))) %>%
  melt(id.vars='Gene')
```

## Boxplot of data
Randomly pick 10 genes for each dataset and check their boxplot
```{r}
sub_gene <- sample(merge_genes, size = 10, replace = FALSE)
sub_lbl <- filter(lbl, Gene %in% sub_gene)
plot(as.factor(sub_lbl$Gene), sub_lbl$value)
```

## range of each dataset
```{r}
range(lbl$value)
range(unc$value)
range(broad$value)
apply(LBL202[1:10,-1],1,median)
apply(LBL202[1:10,-1],1,mean)
apply(Broad202[1:10,-1],1,median)
apply(Broad202[1:10,-1],1,mean)
apply(UNC202[1:10,-1],1,median)
apply(UNC202[1:10,-1],1,mean)
```
After checking the range of gene expression values for each data set, we found out that only UNC202 dataset has negative values, implying that it might be log-transformed already. However, the other two datasets don't contain negative values, so we believed that they are subjective to log-transform. In addition, the three datasets are not mean/median centered either. Therefore, our next step should be log-transforming data and centering data by its median.

## Get Log-transformed & median centered data
```{r}
#lbl$value=log(lbl$value)
range(lbl$value)
#broad$value = log(broad$value)
range(broad$value)
LBL202[,-1] = t(apply(LBL202[,-1],1,function(x) {x - median(x)}))
UNC202[,-1] = t(apply(UNC202[,-1],1,function(x) {x - median(x)}))
Broad202[,-1] = t(apply(Broad202[,-1],1,function(x) {x - median(x)}))
```

## Reshape the data again
```{r}
merge_genes=intersect(LBL202$Gene,UNC202$Gene) %>% intersect(Broad202$Gene)
merge_ids=intersect(colnames(LBL202)[-1],colnames(UNC202)[-1]) %>% 
  intersect(colnames(Broad202)[-1])
unc= UNC202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(UNC202))) %>%
  melt(id.vars="Gene")
broad=Broad202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(Broad202))) %>%
  melt(id.vars='Gene')
lbl=LBL202 %>%
  filter(Gene %in% merge_genes) %>%
  select(Gene,match(merge_ids,colnames(LBL202))) %>%
  melt(id.vars='Gene')
```

## Boxplot of data
Verify that the data has been median-centered and log-transformed
```{r}
sub_gene <- sample(merge_genes, size = 10, replace = FALSE)
sub_lbl <- filter(lbl, Gene %in% sub_gene)
sub_unc <- filter(unc, Gene %in% sub_gene)
sub_broad <- filter(broad, Gene %in% sub_gene)
plot(as.factor(sub_lbl$Gene), sub_lbl$value)
plot(as.factor(sub_unc$Gene), sub_unc$value)
plot(as.factor(sub_broad$Gene), sub_broad$value)
```

## Normalize data
```{r}
total=unc %>%
  merge(broad,by=c('Gene','variable')) %>%
  merge(lbl,by=c('Gene','variable')) %>%
  `colnames<-`(value=c('Gene','Case','unc','broad','lbl'))
totaltry=total %>% 
  mutate(
    unc=scale(unc),
    broad=scale(broad),
    lbl=scale(lbl))
```

## Factor Analysis
```{r}
fit=factanal(totaltry[,3:5],1, scores = "regression")
print(fit)
scores =fit$scores
loadings = fit$loadings
fa = data.matrix(totaltry[,3:5]) %*% loadings
fa_matrix = matrix(fa, ncol = 202, byrow=TRUE)
true_matrix = data.matrix(unifiedScaled[,-1])
```

# Consensus Clustering
Reference: Consensus Clustering: A Resampling-Based Method for Class Discovery and Visualization of Gene Expression Microarray Data ("http://download.springer.com/static/pdf/906/art%253A10.1023%252FA%253A1023949509487.pdf?originUrl=http%3A%2F%2Flink.springer.com%2Farticle%2F10.1023%2FA%3A1023949509487&token2=exp=1489959966~acl=%2Fstatic%2Fpdf%2F906%2Fart%25253A10.1023%25252FA%25253A1023949509487.pdf%3ForiginUrl%3Dhttp%253A%252F%252Flink.springer.com%252Farticle%252F10.1023%252FA%253A1023949509487*~hmac=c5eb7ab0ea30610b34447b1f0f0985f7109d22a58ab9f13702315a47ecd1f53e")

# Install the package "ConsensusClusterPlus"
Instruction is on "https://www.bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html"
```{r}
#source("https://bioconductor.org/biocLite.R")
#biocLite("ConsensusClusterPlus")
library(ConsensusClusterPlus)
```

```{r}
#install.packages('fastcluster')
#library(fastcluster)
dat= read_delim("222_data/unifiedScaledFiltered.txt",
                           "\t", escape_double = FALSE, col_names = F, 
                           trim_ws = TRUE, skip = 1) %>%
    `colnames<-`(value=c('Gene',seq(1,202,by=1)))
#seems best
cluster1=hclust(as.dist(1-cor(dat[,-1],use='everything',method='pearson')))   #linkage=Complete/using euclidean distance
plot(cluster1)
#not good
cluster2=hclust(as.dist(1-cor(dat[,-1],use='everything',method='pearson')),
                method='single') #linkage=Single/using euclidean distance
plot(cluster2)
#used in paper, seems not good as cluster1
cluster3=hclust(as.dist(1-cor(dat[,-1],use='everything',method='pearson')),
                method='average') #linkage=Average/using euclidean distance
pred=cutree(cluster3,4)
plot(cluster3)
#not good
cluster4=hclust(as.dist(1-cor(dat[,-1],use='everything',method='pearson')),
                method='centroid') #linkage="centroid"/using euclidean distance
plot(cluster4)
```

#### Consensus Clustering with hierarchical clustering algorithm
```{r}
library(dplyr)
library(readr)
library(ConsensusClusterPlus)
firstrow=read_delim("222_data/unifiedScaled.txt", 
                    "\t", escape_double = FALSE, col_names = FALSE,
                    trim_ws = TRUE, skip = 0,n_max=1)
dat= read_delim("222_data/unifiedScaledFiltered.txt",
                           "\t", escape_double = FALSE, col_names = F, 
                           trim_ws = TRUE, skip = 1) %>%
    `colnames<-`(value=c('Gene',firstrow))

rownames(dat)=dat$Gene
dat = dat %>%
  select(-Gene)

res = ConsensusClusterPlus(as.matrix(dat),maxK=10,reps=1000,pItem=0.8,pFeature=1
                        ,clusterAlg="hc",distance="pearson",seed=12,plot="png")

labels=as.data.frame(t(cutree(res[[4]]$consensusTree,4))) %>%
  `colnames<-`(colnames(dat))

dat_new=bind_rows(dat,labels)

```

### sigclust analysis
```{r}
#pairwise comparison of clusters using 4 clusters
library(sigclust)
clust_labels <- data.frame(class = res[[4]]$consensusClass,stringsAsFactors =F)
clust_labels <- data.frame(ID = rownames(clust_labels), class = clust_labels$class,stringsAsFactors =F)
#clust_labels=clust_labels[sildist[,3]>0,]
pvals <- matrix(NA, nrow = 4, ncol = 4)

#dat_temp=t(scale(t(dat)))
for(i in 1:3) {
  for(j in (i+1):4) {
    samples <- clust_labels[clust_labels$class == i | clust_labels$class == j,]
    samples$class[samples$class == i] <- 1
    samples$class[samples$class == j] <- 2
    out <- sigclust(x = t(dat[samples$ID]), nsim = 1000, labflag = 1, label = samples$class, icovest = 1)
    pvals[j, i] <- out@pval
  }
}
pvals
```

#### Consensus Clustering with k-means clustering algorithm
```{r}
res_k = ConsensusClusterPlus(as.matrix(dat),maxK=10,reps=1000,pItem=0.8,pFeature=1
                        ,clusterAlg="kmdist",distance="pearson",seed=12,plot="png")
labels_k=as.data.frame(t(cutree(res_k[[4]]$consensusTree,4))) %>%
  `colnames<-`(colnames(dat))

clust_labels_k <- data.frame(class = res_k[[4]]$consensusClass,stringsAsFactors =F)
clust_labels_k <- data.frame(ID = rownames(clust_labels_k), class = clust_labels_k$class,stringsAsFactors =F)
pvals_k <- matrix(NA, nrow = 4, ncol = 4)
dat_temp_k=t(scale(t(dat)))
for(i in 1:3) {
  for(j in (i+1):4) {
    samples_k <- clust_labels_k[clust_labels_k$class == i | clust_labels_k$class == j,]
    samples_k$class[samples_k$class == i] <- 1
    samples_k$class[samples_k$class == j] <- 2
    out_k <- sigclust(x = t(dat_temp_k[,samples_k$ID]), nsim = 1000, labflag = 1, label = samples_k$class, icovest = 1)
    pvals_k[j, i] <- out_k@pval
  }
}
pvals_k
colnames(pvals_k) <- c("Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4")
rownames(pvals_k) <- c("Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4")
```

#### Try core gnes
```{r}
library(cluster)
dmatrix=1-cor(dat)
sildist=silhouette(labels,dmatrix=dmatrix)
```

#SAM: significance anlysis of microarrays
```{r}
#install.packages("samr")
#source("https://bioconductor.org/biocLite.R")
#biocLite("impute")
library(samr)
set.seed(100)
x<-matrix(rnorm(1000*20),ncol=20)
dd<-sample(1:1000,size=100)
u<-matrix(2*rnorm(100),ncol=10,nrow=100)
x[dd,11:20]<-x[dd,11:20]+u
y<-c(rep(1,10),rep(2,10))
samfit<-SAM(x,y,resp.type="Two class unpaired")
# examine significant gene list
print(samfit)
# plot results
plot(samfit)
```
